packages/scrapers-v2/src/runtime/pipeline.ts

```diff
@@ -0,0 +1,98 @@
+import { fetchAndTokenizeHTML } from "../tokenize/tokenize";
+import { addTypeToUrl } from "../classify/classify";
+import { scrapeMajorLinks } from "../urls/urls";
+import { CatalogEntryType, TypedCatalogEntry } from "../classify/types";
+import { Err, Ok, ResultType } from "@graduate/common";
+import { Pipeline, StageLabel } from "./types";
+import { createAgent } from "./axios";
+import { logProgress, logResults } from "./logger";
+
+/**
+ * Runs a full scrape of the catalog, logging the results to the console.
+ * Currently, does nothing with the output. To run from cli, run `yarn scrape`
+ * in `scrapers-v2` dir. Also see `main.ts`.
+ */
+export const runPipeline = async () => {
+  const unregisterAgent = createAgent();
+  const { entries, unfinished } = await scrapeMajorLinks(2021, 2022);
+  if (unfinished.length > 0) {
+    console.log("didn't finish searching some entries", ...unfinished);
+  }
+
+  const pipelines = entries.map((entry) =>
+    createPipeline(entry)
+      .then(addPhase(StageLabel.Classify, addTypeToUrl))
+      .then(
+        addPhase(StageLabel.Filter, filterEntryType, [CatalogEntryType.Major])
+      )
+      .then(addPhase(StageLabel.Tokenize, tokenizeEntry))
+  );
+  const results = await logProgress(pipelines);
+  unregisterAgent();
+  logResults(results);
+};
+
+// convenience constructor for making a pipeline
+const createPipeline = (input: URL): Promise<Pipeline<URL>> => {
+  return Promise.resolve({
+    id: input,
+    trace: [],
+    result: Ok(input),
+  });
+};
+
+/**
+ * Wraps the provided function with a try/catch so that errors don't break the
+ * whole scraper.
+ *
+ * @param phase The identifier for the stage, to be recorded in pipeline trace.
+ * @param next  The function representing this stage. the first argument of this
+ *   function must be the primary entry input.
+ * @param args  Any additional arguments the stage function requires.
+ */
+const addPhase = <Input, Args extends any[], Output>(
```

Comment:
A little confused on how this works, but will come back and have another look at it.